src/main/scala/com/lucidchart/open/relate/ColReader.scala

```diff
@@ -0,0 +1,76 @@
+package com.lucidchart.open.relate
+
+import java.nio.ByteBuffer
+import java.sql.ResultSet
+import java.util.{Date, UUID}
+import shapeless._
+
+class NullColumnException(col: String) extends Exception(s"Unexpected null value in column: $col")
+
+trait ColReader[A] { self =>
+  def read(col: String, row: SqlResult): Option[A]
+
+  def forceRead(col: String, row: SqlResult): A = {
+    read(col, row).getOrElse {
+      throw new NullColumnException(col)
+    }
+  }
+
+  def map[B](f: A => B): ColReader[B] = ColReader[B] { (col, rs) =>
+    self.read(col, rs).map(f)
+  }
+
+  def flatMap[B](f: A => ColReader[B]): ColReader[B] = ColReader[B] { (col, rs) =>
+    self.read(col, rs) match {
+      case Some(a) => f(a).read(col, rs)
+      case None => None
+    }
+  }
+
+  def filter(f: A => Boolean): ColReader[A] =
+    ColReader[A] { (col, rs) => self.read(col, rs).filter(f) }
+}
+
+object ColReader {
+  def apply[A](f: (String, SqlResult) => Option[A]): ColReader[A] = new ColReader[A] {
+    def read(col: String, rs: SqlResult): Option[A] = f(col, rs)
+  }
+
+  def option[A](x: A, rs: ResultSet): Option[A] = {
+    if (x == null || rs.wasNull()) {
```

Comment:
The `rs.wasNull()` check is sufficient.

Note that `x == null` is not sufficient, because `ResultSet#getInt` returns 0 for NULL.